prime-router/docs/proposals/0020-univeral-pipeline-topics.md

```diff
@@ -0,0 +1,63 @@
+# Topic partitioning in the Universal Pipeline
+
+## Introduction
+
+The Universal Pipeline is meant to be the path for all data going through ReportStream. This includes Electronic
+Laboratory Reporting (ELR) as well as Electronic Test Orders and Results (ETOR), which are associated with different
+message types and have different purposes. These two use cases utilize different message types, require different data
+fields to be provided, and will likely be relevant to different sets of senders and receivers.
+
+## Goals
+
+Since ELR and ETOR have different purposes and their data is held to different expectations, we want organization
+settings to prevent the two data types (as well as future other data types) from getting mixed up between
+organizations (senders/receivers) that are intended for the other data type. The solution for this should be simple,
+foolproof, and independent as laid out here:
+
+- Simple: The Requirement is best described with an example: It is likely that ETOR onboarding and operations will be
+  managed by Technical Assistants (TA) hired directly by CDC, not by people on the ReportStream team. Those
+  non-ReportStream teams need a very simple foolproof way to stovepipe their data, to avoid confusion and errors.
+- Foolproof: The implementation should default to not send -- it can't depend on a human remembering to put in an extra
+  clause in one of the filters. (Although that solution may be fine for a UP MVP).
+- Independent of the data itself: You can easily conceive of situations where the data itself might not contain enough
+  info to determine if it belongs within a particular topic. Or, situations where the customers repeatedly change
+  datatypes as they set up their system. Therefore, the information that tells us where certain data goes should be tied
+  to senders and receivers, and may not be solely determinable from the data itself. Topics should be an abstraction.
+
+## Proposal
+
+Currently, the Universal Pipeline (UP) is used for both use cases, and organization settings specify to use the UP by
+setting the `topic` field to `full-elr`. This proposal suggests retaining the `full-elr` topic, but also adding a `etor`
+topic. Those topics will both signify usage of the UP, however senders for one topic will only be able to send data to
+receivers of the same topic. Additionally, each topic may have a different set of default receivers; for instance, the
```

Comment:
each topic may have a different set of default `filters`?

Reply:
`Those topics will both signify usage of the UP`. Regarding this, presently topics are used to determine what pipeline is used, and I think we should work to DECOUPLE the pipeline from the topic. Have we thought about this? See this ticket for reference: https://github.com/CDCgov/prime-reportstream/issues/8592 mainly the following point: `If you decide to resurrect 'topic' and use it as originally structured in the settings, then you'll need some other technique to decide when to route data to new vs old pipelines.`

Reply:
Yea good catch, will make that update.

As for the decoupling, the reality is that it is tightly coupled, topics are implemented/tested/intended for one pipeline and should not be used for the other pipeline, so I don't see much value in decoupling the logic in this case. In the longer term, there won't really be any logic for which pipeline to use since everything will just go to the UP, but for now, we can kind of think of `covid-19` and `monkeypox` as special cases, and the default assumption logically is that the UP is used. Does that make sense? I don't have a specific issue with your proposal of separate api endpoints in that ticket, but it does feel a bit more complicated/disruptive than just having the topic enum include pipeline logic until the legacy pipeline is phased out.

Reply:
I think this is fine for now and can get tackled separately later down the road if needed. Thank you for looking into this!